### PR TITLE
[spruce] Make list results iterable

### PR DESCRIPTION
**Pull request by jhamon**
## Problem

Results from list_indexes a bit annoying to work with because the object isn't iterable

## Solution

Proxy the object and add methods to make it iterable. This gives most of the benefits of drilling into the response and returning the list under the `indexes` while maintaining the flexility to access any attributes that get added to `IndexList` in the future (e.g. for pagination)

## Usage

```python
>>> from pinecone import Pinecone
>>> p = Pinecone(api_key='API_KEY')
>>> results = p.list_indexes()
>>> results.__class__
<class 'pinecone.models.iterable_index_list.IterableIndexList'>
```

Use this object as if it a normal list
```
>>> len(results)
2
>>> results[0]
{'dimension': 3,
 'host': 'jen3-sl-126ib2q.svc.dev-aws-ram-paid.pinecone.io',
 'metric': 'cosine',
 'name': 'jen3-sl',
 'spec': {'serverless': {'cloud': 'aws', 'region': 'us-west-2'}},
 'status': {'ready': True, 'state': 'Ready'}}
>>> names = [x.name for x in results]
>>> names
['jen3-sl', 'jen2-sl']
```

However the representation in the console reflects that it is an object with attributes. And those attributes can still be accessed (e.g. `results.indexes`)
```
>>> results.indexes
[{'dimension': 3,
 'host': 'jen3-sl-126ib2q.svc.dev-aws-ram-paid.pinecone.io',
 'metric': 'cosine',
 'name': 'jen3-sl',
 'spec': {'serverless': {'cloud': 'aws', 'region': 'us-west-2'}},
 'status': {'ready': True, 'state': 'Ready'}}, {'dimension': 2,
 'host': 'jen2-sl-126ib2q.svc.apw5-4e34-81fa.pinecone.io',
 'metric': 'cosine',
 'name': 'jen2-sl',
 'spec': {'serverless': {'cloud': 'aws', 'region': 'us-west-2'}},
 'status': {'ready': False, 'state': 'Initializing'}}]
>>> results
{'indexes': [{'dimension': 3,
              'host': 'jen3-sl-126ib2q.svc.dev-aws-ram-paid.pinecone.io',
              'metric': 'cosine',
              'name': 'jen3-sl',
              'spec': {'serverless': {'cloud': 'aws', 'region': 'us-west-2'}},
              'status': {'ready': True, 'state': 'Ready'}},
             {'dimension': 2,
              'host': 'jen2-sl-126ib2q.svc.apw5-4e34-81fa.pinecone.io',
              'metric': 'cosine',
              'name': 'jen2-sl',
              'spec': {'serverless': {'cloud': 'aws', 'region': 'us-west-2'}},
              'status': {'ready': False, 'state': 'Initializing'}}]}
```

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

## Test Plan

Added unit tests

### CODE REVIEW

**Comment by gdj0nes** [comment]
Can you just update the existing classes with __iter__ ?

**Comment by jhamon** [comment]
> Can you just update the existing classes with `__iter__`? 
> 
> ```
> def __iter__(self):
>     return self.indexes 
> ```
> 
>  

No, because they are outputs of the generator and all models share the same mustache template.

File: tests/unit/models/test_iterable_index_list.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+@pytest.fixture
+def index_list_response():
+    return IndexList(
+            indexes=[
+                {
+                    "name": "test-index-1",
+                    "dimension": 2,
+                    "metric": "cosine",
+                    "capacity_mode": "serverless",
```
Comment (austin-denoble) [inline tests/unit/models/test_iterable_index_list.py:13]: Do these index models need to be updated to the new `spec` shape? Could we use `IndexModel` directly like below in `test_control.py`?

**Review by austin-denoble — APPROVED** [review]
Very nice, wrapping these with some further utilities feels like the right thing to do here. Great work.

File: tests/unit/models/test_iterable_index_list.py (hunk 1 above)
Comment (jhamon) [reply tests/unit/models/test_iterable_index_list.py]: Copilot let me down here 😂 

I'll update them. It doesn't really affect the way the tests work, but it makes the test better documentation if we use correct data.

**Pull request merged**